### DISCUSSION TITLE
AccountInfo data decoding

### DISCUSSION THREAD

**Discussion by tratteo** [bmresearch/Solnet#378] [Q&A] [answered]
Good evening,
how can i decode the `AccountInfo.Data` into a jsonParsed/readable content?

**Comment by hoakbuilds** [comment, marked as answer]
Taken from #140, an old reply.

This depends on the kind of object you're trying to decode but here's a general idea of what to do to try and decode the data:

First and foremost decode the data from base64 into a `byte[]` a `Span<byte>` or whatever your preferred type for this is going to be, you can check some examples of this being done in the [Serum code](https://github.com/bmresearch/Solnet.Serum/tree/master/Solnet.Serum/Models), or the [Mango](https://github.com/bmresearch/Solnet.Mango/tree/master/Solnet.Mango/Models) or [Solend](https://github.com/bmresearch/Solnet.Solend/tree/master/Solnet.Solend/Models) ones
Afterwards, it depends on the type of account you're trying to decode, certain accounts will have a certain data layout, i.e. Serum Market accounts have a certain layout ([Serum Market account layouts](https://github.com/bmresearch/Solnet.Serum/blob/master/Solnet.Serum/Models/Market.cs#L20-L131)), and they have to be decoded according to that layout, by using offsets and reading the desired data type for the account type/layout at the specific offset, as you can see by the static `Deserialize` method on the `Market` class [here](https://github.com/bmresearch/Solnet.Serum/blob/master/Solnet.Serum/Models/Market.cs#L240).

To add more context to this, the v0.3.1 release featured additions of utility methods that can be used to make encoding and decoding of program and account data much easier, the implementations of SPL and Native programs inside `Solnet.Programs` were refactored to use these methods, while the libraries for Serum, Mango, Solend and even the code generation for Anchor programs depends on these methods. You can check those methods here:

[Encoding](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Programs/Utilities/Serialization.cs)
[Decoding](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Programs/Utilities/Deserialization.cs)

They should make the program and account data encoding and decoding much cleaner and easy to implement, they are also open to additions :)

**Reply by tratteo** [reply]
Thanks for the exhaustive answer 👍🏼
Where am i able to find the offset for the different models? In the specific case i am trying to parse the data returned by `GetProgramAccounts`

**Reply by hoakbuilds** [reply]
To find the offsets you will need an IDL (in the case of Anchor programs) or the source code for a rust/ts library, i.e I made the Serum and Mango libraries from looking at the rust implementation and seeing what data types each of the accounts had, and would build the offsets/layouts from there.

**Reply by tratteo** [reply]
Got it. What about the AccountInfo data returned from the Token program (TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA)?

**Reply by hoakbuilds** [reply]
Same logic applies, but we already have those implemented in the library, you can use the RpcClient to get the token account info decoded

**Reply by tratteo** [reply]
Yes, i see i can call `GetTokenAccountInfo` in order to obtain the `ParsedTokenAccountInfo`, however this would mean calling another RPC just to get the parsed data. In my case i just need the amount of the token and i could get it from the raw data instead of executing another RPC.

**Reply by hoakbuilds** [reply]
But if you have the raw data that means you already did a call to the RPC, if you did a `GetAccountInfo` you can replace that for the `GetTokenAccountInfo`. 

Anyway, you have the structures implemented [here](https://github.com/bmresearch/Solnet/blob/master/src/Solnet.Rpc/Models/AccountData.cs) but I believe we request the data as `jsonParsed` instead of the raw base64. If you want to decode the raw base64 you will need to see the rust structure and decode the appropiate data type at the offset of the token amount. The [offset is 64 and you will want to decode a `ulong`](https://github.com/solana-labs/solana-program-library/blob/master/token/program/src/state.rs#L92).

**Reply by tratteo** [reply]
That is exactly what i was looking for.
Thanks a lot for the time and fast responses!
